### DISCUSSION TITLE
Airflow can't extract host and port in Connection uri that has multiple hosts

### DISCUSSION THREAD

**Discussion by boushphong** [apache/airflow#35893] [General] [answered]
### Apache Airflow version

2.7.3

### What happened

We have a Secret backend that retrieves connection uri. We encountered an error where if we have a connection uri that has multiple hosts (like a MongoDB sharded cluster uri). The host and port parts will be extracted incorrectly and cause an error:
```
  File "/Users/buiducphong/Projects/airflow/venv/lib/python3.10/site-packages/sqlalchemy/orm/state.py", line 481, in _initialize_instance
    with util.safe_reraise():
  File "/Users/buiducphong/Projects/airflow/venv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/Users/buiducphong/Projects/airflow/venv/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/Users/buiducphong/Projects/airflow/venv/lib/python3.10/site-packages/sqlalchemy/orm/state.py", line 479, in _initialize_instance
    return manager.original_init(*mixed[1:], **kwargs)
  File "/Users/buiducphong/Projects/airflow/venv/lib/python3.10/site-packages/airflow/models/connection.py", line 127, in __init__
    self._parse_from_uri(uri)
  File "/Users/buiducphong/Projects/airflow/venv/lib/python3.10/site-packages/airflow/models/connection.py", line 199, in _parse_from_uri
    self.port = uri_parts.port
  File "/opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/parse.py", line 185, in port
    raise ValueError(f"Port could not be cast to integer value as {port!r}")
ValueError: Port could not be cast to integer value as '20019,dummyhost2:20019'
```

The full connection uri is `mongodb://dummy_user:dummy_password@dummyhost1:20019,dummyhost2:20019/`, however when Airflow Connection evaluates the string, it will extract `dummyhost1` as the host and `20019,dummyhost2:20019` as the port which would lead to a `ValueError`

### What you think should happen instead

Airflow Connection should be able handle the case where the host consists of multiple hosts by default.

### How to reproduce

Run this code:

```python3
from airflow.models import Connection

conn = Connection(
    conn_id="dummy_connection", uri="mongodb://dummy_user:dummy_password@dummyhost1:20019,dummyhost2:20019/"
)
```

### Operating System

Ubuntu

### Deployment

Other

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
Currently, MongoHook doesn't support multiple hosts, but it's possible to support it through the extras, not the host, where your connection is not a valid URI. But it will be much easer to achieve that with a new extra, where we can check if it's present to use it, and if not, we fallback to the connection host.

cc: @potiuk @Taragolis WDYT?

**Reply by potiuk** [reply, marked as answer]
I think regardless of the solution how to keep it in connection, the problem in this case is that you are passing "mongodb://" URL in "airlfow" URL - these two are different. This is what @Taragolis started to discussed in https://lists.apache.org/thread/3n832wdy8xf2ochw0s7kmys23vxmwd8j  - this is utterly confusing that airflow's URL is different than the URL target connection should use.

I think it's ok to use mutliple hosts in "host" of the connection and adding it to MongoDBHook (but someone will have to do it) - but even then, the actual URL to set in your secret will be different than the one you specified @boushphong You will need to generate your URL using method described here: https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#generating-a-connection-uri
